akka-typed/src/main/scala/akka/typed/internal/ActorCell.scala

```diff
@@ -0,0 +1,433 @@
+          unsafe.getAndAddInt(this, statusOffset, -1)
+        }
+      } else if (Debug) println(s"[$thread] $self NOT enqueueing $signal: terminated")
+    } catch handleException
+  }
+
+  /**
+   * Main entry point into the actor: the ActorCell is a Runnable that is
+   * enqueued in its Executor whenever it needs to run. The _status field is
+   * used for coordination such that it is never enqueued more than once at
+   * any given time, because that would break the Actor Model.
+   *
+   * The idea here is to process at most as many messages as were in queued
+   * upon entry of this method, interleaving each normal message with the
+   * processing of all system messages that may have accumulated in the
+   * meantime. If at the end of the processing messages remain in the queue
+   * then this cell is rescheduled.
+   *
+   * All coordination occurs via a single Int field that is only updated in
+   * wait-free manner (LOCK XADD via unsafe.getAndAddInt), where conflicts are
+   * resolved by compensating actions. For a description of the bit usage see
+   * the companion object’s source code.
+   */
+  override final def run(): Unit = {
+    if (Debug) println(s"[$thread] $self entering run(): interrupted=${Thread.currentThread.isInterrupted}")
+    val status = _status
+    val msgs = messageCount(status)
+    var processed = 0
+    try {
+      unscheduleReceiveTimeout()
+      if (!isClosed(status)) {
+        while (processAllSystemMessages() && !queue.isEmpty() && processed < msgs) {
+          val msg = queue.poll()
+          processed += 1
+          processMessage(msg)
+        }
+      }
+      scheduleReceiveTimeout()
+    } catch {
+      case NonFatal(ex) ⇒ fail(ex)
+      case ie: InterruptedException ⇒
+        fail(ie)
+        if (Debug) println(s"[$thread] $self interrupting due to catching InterruptedException")
+        Thread.currentThread.interrupt()
+    } finally {
+      // also remove the general activation token
+      processed += 1
+      val prev = unsafe.getAndAddInt(this, statusOffset, -processed)
+      val now = prev - processed
+      if (isClosed(now)) {
+        // we’re finished
+      } else if (activations(now) > 0) {
+        // normal messages pending: reverse the deactivation
+        unsafe.getAndAddInt(this, statusOffset, 1)
+        // ... and reschedule
+        executionContext.execute(this)
+      } else if (_systemQueue.head != null) {
+        /*
+         * System message was enqueued after our last processing, we now need to
+         * race against the other party because the enqueue might have happened
+         * before the deactivation (above) and hence not scheduled.
+         *
+         * If we win, we reschedule; if we lose, we must remove the attempted
+         * activation token again.
+         */
+        val again = unsafe.getAndAddInt(this, statusOffset, 1)
+        if (activations(now) == 0) executionContext.execute(this)
```

Comment:
isn't now a val calculated at L291? Also, we checked that this can't be > 0, and also that is not negative (isClosed would be bit 31 right?). So how can this be not zero here? Didn't you want to use `again` here?